lib/Test2/Formatter/TeamCity.pm

```diff
@@ -127,35 +220,92 @@ sub _start_suite {
         );
     }
 
+    my $realtime = _not_empty($name)
+        && ( !$self->{ +_CURRENT } || $self->{ +_CURRENT }->realtime );
+    if ($realtime) {
+        _debug('-> which will be outputting in real time');
+    }
+    else {
+        _debug('-> which will not be outputting in real time');
+    }
+
     my $suite = Test2::Formatter::TeamCity::Suite->new(
-        id     => $id,
-        name   => $name,
-        parent => $self->{ +_CURRENT },
+        id       => $id,
+        name     => $name,
+        parent   => $self->{ +_CURRENT },
+        realtime => $realtime,
     );
 
     if ( $self->{ +_CURRENT } ) {
-        $self->{ +_CURRENT }->add_child($suite);
+        if ( $self->{ +_CURRENT }->realtime ) {
+
+            # flush out any events we've been building up in our parent
+            # since we're starting a new subtest
+            $self->_flush_current_children_events;
+        }
+        else {
+            # record this subtest (we don't need to do this if we're
+            # outputting in realtime because by the time we're done
+            # we'll have already output everything
+            $self->{ +_CURRENT }->add_child($suite);
+        }
     }
 
     $self->{ +_TOP } //= $suite;
     $self->{ +_CURRENT } = $suite;
 
+    if ( $suite->realtime ) {
+        $self->_tc_message(
+            testSuiteStarted => {
+                name => $name,
+            }
+        );
+    }
+
     return;
 }
 
 sub _finish_suite {
     my $self = shift;
     my $id   = shift;
 
+    _debug( 'finish suite: ' . $id );
+
     unless ( $self->{ +_CURRENT } ) {
         warn 'Called _finish_suite before any suites were started';
         return;
     }
 
-    my $last_suite = $self->{ +_CURRENT }{id};
-    unless ( $last_suite eq $id ) {
+    while (1) {
+        my $last_suite = $self->{ +_CURRENT }{id};
+        last if $last_suite eq $id;
+
         warn
             "Last suite on the stack ($last_suite) does not match suite we want to finish ($id)";
+
+        $self->_finish_suite($last_suite);
+    }
+
+    if ( $self->{ +_CURRENT }->realtime ) {
+
+        # if we're running in realtime then we need to output any remaining
+        # child events and finish the suite *right* *now*
+
+        # flush out any events still in the queue.
+        $self->_children_to_tc( $self->{ +_CURRENT }->children );
+
+        $self->_tc_message(
+            testSuiteFinished => {
+                name => $self->{ +_CURRENT }->name,
+            }
+        );
+    }
+    elsif ($self->{ +_CURRENT }->parent
+        && $self->{ +_CURRENT }->parent->realtime ) {
```

Comment:
As per our discussion, this should probably check that all parents are in realtime mode, not just the direct parent